### ISSUE TITLE
[cat] read from stdin

### ISSUE THREAD

**Issue by rufuspollock** [datopian/data-cli#169]
As a Wrangler I want to pass a stream of data to data cat and have it work so that I can use my favourite tool or system for opening a data stream and take advantage of data cat's presentation system (e.g. i may need to pass credentials or use curl's amazing features in opening a stream)

The following should work fine:

```bash
# results in same as data cat mydata.csv
cat mydata.csv | data cat

# results in same as data cat url
curl mydata.csv | data cat
```

We can also make the source explicit using `_` to stand for stdin so that we don't end up with ambiguity ...:

```bash
# read from mydata.csv and write to out.xls excel file
cat mydata.csv | data cat _ out.xls
cat mydata | data cat _.csv out.xls # explicitly hint on the input format
cat mydata | data cat _.csv # write to stdout as per default ...
```

## Acceptance criteria

* [x] The examples above work

## Tasks

* [x] Update parsing for input "paths" to support _ as a special case signifying stdin
  * [ ] Note a bonus that we should also use _ for stdout for writing for dumpers so we can do `data cat xyz.csv _.xls` (writes xls to stdout)
* [x] Update to support special case where no input or output path are specified

## Analysis

We need to read from stdin if `process.stdin` available.

> The process.stdin property returns a stream connected to stdin (fd 0). It is a net.Socket (which is a Duplex stream) unless fd 0 refers to a file, in which case it is a Readable stream.

Check if stdin is a Socket and if so listen for readable data:

```javascript
// also parse provided input path so we can check if user wants to pipe a stream
const pathParts = path.parse(argv._[0])

let pipedData
if (process.stdin instanceof Socket && pathParts.name === '_') {
  process.stdin.on('data', () => {
    pipedData += process.stdin.read()
  })

  process.stdin.on('end', () => {
    dump(pipedData)
  })
} else if (pathParts.name === '_') {
  info('No stream provided.')
} else {
  const res = File.load(argv._[0])
  if (argv._[1]) {
    outFileExt = path.extname(argv._[1])
  }
  dump(res)
}

const dump = async (res) => {
  if (!argv._[1] || argv._[1] === 'stdout') {
    const out = await dumpers.ascii(res)
    console.log(out)
  } else if (outFileExt === '.csv') {
    const out = await dumpers.csv(res)
    fs.writeFileSync(argv._[1], out)
    console.log(`Your data is saved in ${argv._[1]}`)
  } else if (outFileExt === '.md') {
    const out = await dumpers.md(res)
    fs.writeFileSync(argv._[1], out)
    console.log(`Your data is saved in ${argv._[1]}`)
  } else if (outFileExt === '.xlsx') {
    const out = await dumpers.xlsx(res)
    const wb = {SheetNames: ['sheet'], Sheets: {sheet: out}}
    XLSX.writeFile(wb, argv._[1])
    console.log(`Your data is saved in ${argv._[1]}`)
  } else {
    console.log('We currently do not support this feature.')
  }
}
```

Now we call dumpers with either file object or data read from stdin. We have to refactor the dumpers to expect the second case:

We can have small function that handles getting rows for all dumpers:

```javascript
const getRows = async (input) => {
  let rows
  if (input.constructor.name === 'File') {
    rows = await toArray(await input.rows())
  } else {
    rows = input // convert to array of arrays
  }
  return rows
}
```

so, for example, ascii dumper would look like:

```javascript
const dumpAscii = async function (input) {
  const rows = await getRows(input)
  ...
}
```

**Comment by anuveyatsu** [comment]
@rufuspollock can you please clarify to me about the following task:

> Note a bonus that we should also use _ for stdout for writing for dumpers so we can do data cat xyz.csv _.xls (writes xls to stdout)

I don't get how we write xls to stdout - how would it look like?

**Comment by rufuspollock** [comment]
@anuveyatsu can you summarize where we are with this issue.

**Comment by rufuspollock** [comment]
@anuveyatsu please can you summarize where we are with this issue 😄

**Comment by anuveyatsu** [comment]
FIXED, will be available in the next release of the CLI tool.
